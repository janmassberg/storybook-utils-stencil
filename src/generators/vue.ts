import {
    ArgsType,
    filterHtmlAttributes,
    filterEventHandlers,
    filterJsxProperties,
    formatHtml,
    formatJsx,
    objectToString,
    toKebabCase,
    toLowerCamelCase,
    ucFirst,
} from "../utils";

const transformVueAttributes = (attributes: ArgsType) => {
    return attributes.map(([key, value]) => {
        return `${toKebabCase(key)}="${value}"`;
    });
};

export const generateSourceCodeVue = (component: string, args: any) => {
    const attribs = filterHtmlAttributes(args);
    const props = filterJsxProperties(args);
    const events = filterEventHandlers(args);
    const slot = args._children || "";

    const vueComponentName = toKebabCase(component);
    const vueAttributes = transformVueAttributes(attribs);
    const vueProps = props.map(
        ([key]) => `.${toLowerCamelCase(key)}="${toLowerCamelCase(key)}"\n`
    );
    const vueEvents = events.map(([key]) => `@${key}="on${ucFirst(key)}"\n`);
    const vueComponentJsx = `<${vueComponentName}
        ${vueAttributes.join("\n")}
        ${vueProps.join("\n")}
        ${vueEvents.join("\n")}
        >${slot}</${vueComponentName}>`;

    const vueEventHandlers = events.map(
        ([key, value]) =>
            `on${ucFirst(key)}: ${value.toString().replace(/\n/g, "\n")}`
    );

    const vueConstants = props.map(([key, value]) => {
        return `${toLowerCamelCase(key)}: ${objectToString(value).replace(
            /\n/g,
            "\n"
        )}`;
    });

    const vueComponentData =
        vueConstants.length < 1
            ? ""
            : `
  data() {
    return {
      ${vueConstants.join(",\n")}
    };
  },`;

    const vueComponentMethods =
        vueEventHandlers.length < 1
            ? ""
            : `
  methods: {
    ${vueEventHandlers.join(",\n")}
  },`;

    const vueComponentScriptSrc = `
export default {
${vueComponentData.replace(/^\n(.*)/, "$1")}
${vueComponentMethods.replace(/^\n(.*)/, "$1")}
};
`;

    const vueComponentSource = `
<template>
  ${vueComponentJsx}
</template>

<script lang="ts">
${formatJsx(vueComponentScriptSrc.replace(/^\n/, ""))}
</script>
`;
    // return vueComponentSource;
    return formatHtml(vueComponentSource);
};
