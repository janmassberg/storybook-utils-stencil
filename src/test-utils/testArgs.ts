export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: (): void => {
        // Implementation
    },
    onTestEvent: (event: Event): void => {
        // Handle event
        console.log(event.type);
    },
    _slot: "Test children",
};

export const testArgsButton = {
    required: true,
    variant: "primary",
    onClick: (event: Event) => {
        // Handle click event
        console.log(event.target);
    },
    _slot: "Label",
};

export const testArgsComponentGenerator = {
    component: "my-component",
    args: testArgs,
};

export const testCopmonentSlot = `<my-child-component foo-bar="foo">
<my-nested-child-component bar-foo="bar"><span class="test-class">Label</span></my-nested-child-component>
</my-child-component>`;
