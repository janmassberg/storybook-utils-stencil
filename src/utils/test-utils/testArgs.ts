export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {},
    onTestEvent: () => {},
    _children: "Test children",
};
