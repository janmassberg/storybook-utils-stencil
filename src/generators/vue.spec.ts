import { generateSourceCodeVue } from "./vue";
import { testArgsComponentGenerator, testCopmonentSlot } from "../test-utils/testArgs";

describe("generateSourceCodeVue", () => {
    it("should generate valid formatted vue JSX code with args", () => {
        expect(
            generateSourceCodeVue(testArgsComponentGenerator)
        ).toMatchSnapshot();
    });

    it("should generate valid formatted vue JSX code with children", () => {
        expect(
            generateSourceCodeVue({
                ...testArgsComponentGenerator,
                args: {
                    _slot: testCopmonentSlot,
                }
            })
        ).toMatchSnapshot();
    });
});
