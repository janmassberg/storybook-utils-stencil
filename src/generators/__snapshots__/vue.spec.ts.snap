// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSourceCodeVue should generate valid formatted vue JSX code with children 1`] = `
"<template>
  <my-component>
    <my-child-component foo-bar="foo">
      <my-nested-child-component bar-foo="bar">
        <span class="test-class">Test children</span>
      </my-nested-child-component>
    </my-child-component>
  </my-component>
</template>

<script lang="ts">
  export default {};
</script>
"
`;

exports[`generateSourceCodeVue should generate valid formatted vue JSX code with props 1`] = `
"<template>
  <my-component
    test-false="false"
    test-true="true"
    test-string="test-string"
    test-number="12345"
    .testArray="testArray"
    .testObject="testObject"
    .testFunction="testFunction"
    @testEvent="onTestEvent"
  >
    Test children
  </my-component>
</template>

<script lang="ts">
  export default {
    data() {
      return {
        testArray: [1, 2, 3],
        testObject: { foo: "bar" },
        testFunction: () => {
          // do nothing
        },
      };
    },
    methods: {
      onTestEvent: (event) => {
        // Handle event
        console.log(event.type);
      },
    },
  };
</script>
"
`;
